### ISSUE TITLE
Behaviour is different when there are no shared groups in TLS1.2 and TLS1.3

### ISSUE THREAD

**Issue by fwh-dc** [openssl/openssl#21256]
OpenSSL 3.1.1 30 May 2023 (Library: OpenSSL 3.1.1 30 May 2023)

Is it intentional that the behaviour when the client and server has no shared groups is different between TLS1.2 and TLS1.3?

If there are no shared groups in TLS1.3 I get a handshake error on the client side and a key share error on the server side.

If there are no shared groups in TLS1.2 the connection is successfully established. I would expect the same behaviour as TLS1.3.

Here are the commands used:
TLS1.2:
```
.local/bin/openssl s_client -curves secp384r1  -provider default -CAfile rsa_CA.crt -tls1_2
.local/bin/openssl s_server -cert rsa_srv.crt -key rsa_srv.key -www -tls1_2 -groups secp521r1 -provider default
```
TLS1.3:
```
.local/bin/openssl s_client -curves secp384r1  -provider default -CAfile rsa_CA.crt -tls1_3
.local/bin/openssl s_server -cert rsa_srv.crt -key rsa_srv.key -www -tls1_3 -groups secp521r1 -provider default
```

**Comment by mattcaswell** [comment]
TLSv1.2 and TLSv1.3 are very different.

With TLSv1.3 the key share gets sent as part of the ClientHello. There must be an overlap in the supported groups for this to work. Otherwise no key share will be acceptable and a handshake failure is the inevitable result.

With TLSv1.2 it is perfectly possible for a handshake to complete without an overlapping set of supported groups if a ciphersuite is negotiated that doesn't need one - for example if a ciphersuite is used that has no forward secrecy, or if a DHE based ciphersuite is used (which doesn't use the supported_groups mechanism in TLSv1.2).

In other words the TLSv1.3 protocol requires there to be shared groups between the client and server, but the TLSv1.2 protocol does not.

> Is it intentional that the behaviour when the client and server has no shared groups is different between TLS1.2 and TLS1.3?

Due to the reasons above: yes.

**Comment by fwh-dc** [comment]
That explains it, thanks for clarifying.

For future reference:
By forcing an ECDHE I get the failure I am expecting:
```
./apps/openssl s_server -cert rsa_srv.crt -key rsa_srv.key -www -tls1_2 -groups secp521r1
./apps/openssl s_client -tls1_2 -groups secp384r1 -cipher ECDHE-RSA-AES256-GCM-SHA384
```
